### PR TITLE
Refactored support for node local dns.

### PR DESCRIPTION
**Pull request by ScheererJ**
Removed feature flag and introduced a shoot specific annotation instead.
All (dns related) network policies now always allow the traffic to the node local dns endpoints.
Node local dns should no longer have a cascading effect requiring shoots to enable it if it is enabled on the seed.

**How to categorize this PR?**

/area networking
/kind api-change

**What this PR does / why we need it**:
Removed feature flag and introduced a shoot specific annotation instead.
All (dns related) network policies now always allow the traffic to the node local dns endpoints.
Node local dns should no longer have a cascading effect requiring shoots to enable it if it is enabled on the seed.

**Which issue(s) this PR fixes**:
None in particular, but it removes the somewhat broken feature gate NodeLocalDNS, which could not be easily (without downtime and additional effort) be switched.

**Special notes for your reviewer**:

**Release note**:

```breaking operator
Gardenlet feature gate NodeLocalDNS was removed and replaced by a shoot specific annotation.
```

### CODE REVIEW

File: charts/shoot-core/components/charts/network-policies/templates/allow-to-dns.yaml [hunk 1]
```diff
@@ -20,10 +20,8 @@ spec:
     - podSelector:
         matchExpressions:
         - {key: k8s-app, operator: In, values: [kube-dns]}
-    {{- if .Values.nodeLocalDNS.enabled }}
     - ipBlock:
         cidr: {{ .Values.nodeLocalDNS.kubeDNSClusterIP }}/32
-    {{- end }}
```
Comment (rfranzke) [inline charts/shoot-core/components/charts/network-policies/templates/allow-to-dns.yaml]: In the shoot, why is this removed? The network policy is only required when the shoot enables the node-local-dns, isn't it?

Comment (ScheererJ) [reply charts/shoot-core/components/charts/network-policies/templates/allow-to-dns.yaml]: As discussed with you and @DockToFuture , I adapted all three network policies (garden namespace in seed, control plane(s) in seed, kube-system in shoot) to always allow the node local dns entries, i.e. remove the if checks.
You can argue though that while it is required to remove them from the network policies in the garden namespace and the control plane(s) unless you want to live with cascading effects, it is possible to let this if block stay in the kube-system network policy as this should not affect anything apart from the cluster itself.
I think removing the node local checks on all three network policies is more consistent, but this is not a hard opinion. So if you insist, I can re-add the if check here.

Comment (rfranzke) [reply charts/shoot-core/components/charts/network-policies/templates/allow-to-dns.yaml]: To my understanding, we only discussed about always allowing both CIDRs for the network policies created by gardenlet's seed controller due to the fact that we don't know which DNS component (coredns/node-local-dns) is running in the seed. For the shoot, however, we definitely know it because it's configured in the `Shoot` resource. Hence, I would indeed only allow required network traffic if we have the possibility to shrink it down.

Comment (ScheererJ) [reply charts/shoot-core/components/charts/network-policies/templates/allow-to-dns.yaml]: Ok, I reverted the changes related to the network policy in the kube-system namespace.
As a positive side effect, it reduces the amount of files being changed.

**Review by rfranzke — APPROVED** [review]
Thanks!
/lgtm

**Review by DockToFuture — APPROVED** [review]
/lgtm

**Comment by danielfoehrKn** [comment]
Should that not have been announced as a breaking change?  @rfranzke @ScheererJ 
Currently, it is not as prominent in the `Others` section and could more easily be overlooked.

**Comment by rfranzke** [comment]
@danielfoehrKn Yes, +1, nice catch

**Comment by ScheererJ** [comment]
> Should that not have been announced as a breaking change? @rfranzke @ScheererJ
> Currently, it is not as prominent in the `Others` section and could more easily be overlooked.

Good point. How would I do that?

=> I changed `others` to `breaking` in the release note header. Is that sufficient?
